### ISSUE TITLE
Why does Body field shows as attributes?

### ISSUE THREAD

**Issue by smoms** [open-telemetry/opentelemetry-dotnet#5361]
I am not sure if it is a bug.

I need to send Log attributes and I am following the apporach described [here](https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/docs/logs/getting-started-aspnetcore/README.md).

This is my log:

```
       [LoggerMessage(LogLevel.Information, "{body}")]
        public static partial void LogInformation(this ILogger logger,
            LogType logType,
            string? body = null,
            string? businessTxKeyIoTDeviceId = null,
            string? businessTxKeyIoTDeviceSn = null,
            string? businessTxKeyOrganizationId = null,
            string? businessTxKeyUserId = null,
            string? businessTxKeyUserGuid = null,
            string? auditEventAction = null,
            string? entityType = null,
            string? entityId = null,
            string? subjectId = null);
```

I get on an OTel Collector system the following log:

```
  "body": "this is my payload message",
  "traceid": "dd61ef6729320698c1d093c0f6674a8b",
  "spanid": "765639bbf2b1514a",
  "severity": "Information",
  "flags": 1,
  "attributes": {
    "body": "this is my payload message",
    "businessTxKeyIoTDeviceSn": "blue_sky",
    "businessTxKeyUserGuid": "12345",
    "businessTxKeyUserId": "abc",
    "logType": "Diagnostic",
    "{OriginalFormat}": "{body}"
```

Questions are:

1. can I avoid to get the body in the attrbiutes section? It is not an attribute for my requirements..
2. how can I remove the "{OriginalFormat}": "{body}" line?

On the other hand, if I do not set 'body' as an attribute there is no way I can find to pass a message/body to the log calls from a client application.. 

thanks

**Comment by cijothomas** [comment]
> can I avoid to get the body in the attrbiutes section? It is not an attribute for my requirements..

Because you have an attribute named "body"!
`string? body = null,`

**Comment by smoms** [comment]
@cijothomas thanks for your answer.
That's right I use body as attrobute.
But again my question, assuming I use this list of attributes (without body):


```
 [LoggerMessage(LogLevel.Information, "{???}")]
        public static partial void LogInformation(this ILogger logger,
            LogType logType,
            string? businessTxKeyIoTDeviceId = null,
            string? businessTxKeyIoTDeviceSn = null,
            string? businessTxKeyOrganizationId = null,
            string? businessTxKeyUserId = null,
            string? businessTxKeyUserGuid = null,
            string? auditEventAction = null,
            string? entityType = null,
            string? entityId = null,
            string? subjectId = null);
```

How can I now send a log message with **body plus attrbiutes**?

**Comment by smoms** [comment]
Actually I have used a Processor to strip the body from the attriburtes and it is what I want to see now:

```
body": "this is my payload message",
  "traceid": "18d9f28764f0753ed00f00ddedb5a37b",
  "spanid": "3a6ab85a68dfe41d",
  "severity": "Information",
  "flags": 1,
  "attributes": {
    "businessTxKeyIoTDeviceSn": "blue_sky",
    "businessTxKeyUserGuid": "12345",
    "businessTxKeyUserId": "abc",
    "logType": "Diagnostic",
    "{OriginalFormat}": "{body}"
  },
```

question is: is this the expected/right approach?

**Comment by cijothomas** [comment]
> @cijothomas thanks for your answer. That's right I use body as attribute. But again my question, assuming I use this list of attributes (without body now):
> 
> ```
>  [LoggerMessage(LogLevel.Information, "{???}")]
>         public static partial void LogInformation(this ILogger logger,
>             LogType logType,
>             string? businessTxKeyIoTDeviceId = null,
>             string? businessTxKeyIoTDeviceSn = null,
>             string? businessTxKeyOrganizationId = null,
>             string? businessTxKeyUserId = null,
>             string? businessTxKeyUserGuid = null,
>             string? auditEventAction = null,
>             string? entityType = null,
>             string? entityId = null,
>             string? subjectId = null);
> ```
> 
> How can I now send a log message with **body plus (the above) attributes**?

isn't that the default behavior you are observing? i.e body + all the above as attributes?
Can you show the full code, so we can take a look at if OTLP Exporter is missing some special casing..

**Comment by smoms** [comment]
I am not sure my questin is clear then.

I need both **attributes** and **body** but not as an attrbiute.
If - as you suggest I remove the body from the attribute list, can you please tell me how I can send the body?

So far I was using body as attribute, which again I do not want as an attribute, and I was calling the looging API with something similar to:


```
logger.LogInformation(LogType.Diagnostic,
                body: "this is my payload message",
                businessTxKeyUserId: "xxx",
                businessTxKeyIoTDeviceId: "xxx",
                businessTxKeyUserGuid: userGuid,          
                businessTxKeyDeviceSn: "12345");
```

Here the configuration:

```
 internal static void ConfigureLoggingOpenTelemetryTraces(this ILoggingBuilder loggingBuilder, TelemetryOptions telemetryOptions)
    {
        loggingBuilder.AddOpenTelemetry(options =>
        {
            options
                .SetResourceBuilder(
                    ResourceBuilder
                        .CreateDefault()
                        .AddService(telemetryOptions.ServiceName, telemetryOptions.ServiceNamespace, telemetryOptions.ServiceVersion));
            if (telemetryOptions.EndpointLog != null)
            {
                options.AddOtlpExporter(o =>
                {
                    o.Protocol = OtlpExportProtocol.HttpProtobuf;
                    o.Endpoint = telemetryOptions.EndpointLog;
                    o.HttpClientFactory = () => CreateHttpHandler();
                });
            }
            else
            {
                options.AddConsoleExporter();
            }
 
        });
    }
```

**Comment by cijothomas** [comment]
I used this example with OTLP Exporter
https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/docs/logs/getting-started-console/Program.cs#L19

And this is what I see in OTel Collector:

```txt
LogRecord #0
ObservedTimestamp: 2024-02-21 14:59:26.8189406 +0000 UTC
Timestamp: 2024-02-21 14:59:26.8189406 +0000 UTC
SeverityText: Information
SeverityNumber: Info(9)
**Body: Str(Food `{name}` price changed to `{price}`.)**
Attributes:
     -> name: Str(artichoke)
     -> price: Double(9.99)
Trace ID:
Span ID:
Flags: 0
LogRecord #1
ObservedTimestamp: 2024-02-21 14:59:26.8479572 +0000 UTC
Timestamp: 2024-02-21 14:59:26.8479572 +0000 UTC
SeverityText: Critical
SeverityNumber: Fatal(21)
Body: Str(A `{productType}` recall notice was published for `{brandName} {productDescription}` produced by `{companyName}` ({recallReasonDescription}).)
Attributes:
     -> brandName: Str(Contoso)
     -> productDescription: Str(Salads)
     -> productType: Str(Food & Beverages)
     -> recallReasonDescription: Str(due to a possible health risk from Listeria monocytogenes)
     -> companyName: Str(Contoso Fresh Vegetables, Inc.)
Trace ID:
Span ID:
```

Both body and attributes are populated correctly. Can you use this example as-is, and see if it matches your expectation? And then adjust your question based on that, to make sure I am following your scenario fully.

**Comment by smoms** [comment]
@cijothomas thanks.
I understand this example. And it is also showing the point of my issue.
Please see the "pre-defined" body field:

`Body: Str(A `{productType}` recall notice was published for `{brandName} {productDescription}` produced by `{companyName}` ({recallReasonDescription}).)`

generate by the code:

```
 [LoggerMessage(LogLevel.Critical, "**A `{productType}` recall notice was published for `{brandName} {productDescription}` produced by `{companyName}` ({recallReasonDescription}).**")]
    public static partial void FoodRecallNotice(
        this ILogger logger,
.....
```

Again questions are:

1. how can a client - using `logger.Informatio(..) ` function - pass a body field **dynamically** instead of having it as a pre-defined string as it is above?
In other words, the body here can just interpolate attributes in a pre-defined string. But what if I want a completely indipendent and dynamic body field that a client can pass in?

Normally in a` Ilogger.Information() ` call a client is free to pass any body/message information freely and dynamically without being constrained by a pre-defined, interpolated string template.

**Comment by cijothomas** [comment]
There is no need to keep using the same body. That is what you'll get by following the source-generation way of using ILogger.

You can use ILogger.Log, pass any attributes, and provide a formatter which can be used to produce "body" dynamically.
To do that:
1. Enable this https://github.com/open-telemetry/opentelemetry-dotnet/tree/main/docs/logs/customizing-the-sdk#includeformattedmessage
2. use the [foundational Log api](https://learn.microsoft.com/en-us/dotnet/api/microsoft.extensions.logging.ilogger.log?view=dotnet-plat-ext-8.0) and pass the formatter where you can control what the body would look like.

let me know if this helps. I can write an example here, if not clear.

**Comment by smoms** [comment]
@cijothomas it would be great an example. many thanks.

**Comment by cijothomas** [comment]
> @cijothomas it would be great an example. many thanks.

```
logger.Log(logLevel: LogLevel.Information,
            eventId: default,
            state: new List<KeyValuePair<string, object>>()
            {
                new KeyValuePair<string, object>("food", "apple"),
                new KeyValuePair<string, object>("price", 3.99)
            },
            exception: null,
            formatter: (state, ex) => "Use any message here.!"
            );
```

"Use any message here.!" will be the body. And you are free to decide what its value will be. I just hard coded it for example.

**Comment by smoms** [comment]
Ok that's is the standard ILogger API. Since my aim is to wrap the log as a library to encapsulate a pre-defined list of attributes here instead there would be a free list of key-value pairs...

I am following the previous apporach where my trick is to have the body as an attrbute and then I strip it out with an OTel Process. I am not sure how much artificial is this ;)

**Comment by cijothomas** [comment]
Closing as there is no further questions left.
